### PR TITLE
Appveyor

### PR DESCRIPTION
**Pull request by DinCahill**
This allows [Appveyor](https://appveyor.com) to build playd in the Release configuration on Win32 and x64. 

### WindowsBuilder.ps1

The main gubbins is WindowsBuilder.ps1, a PowerShell script for automatically downloading and building all of playd's dependencies, and building playd. I still need to write a help message, so here's how to use it:

`.\WindowsBuilder.ps1 -arch (x86|x64) [-deps] [-playd]`

#### Requirements
- ~~The `-sh` switch needs a path to Cygwin bash. Cygwin needs a C compiler, automake, and autotools. It is used for compiling the [gendef](https://sourceforge.net/p/mingw-w64/wiki2/gendef) tool from MinGW-w64, in order to make a def file for libmpg123-0.dll.~~ No longer used.
- Python 2.7 is required for building gyp for building libuv. Put it in your `PATH`, or set `PYTHON` to the path of your Python exe. In PS: `$env:PYTHON = "path\to\python.exe"`

Right now, it builds both deps and playd if both switches are omitted.

All files generated are put in a directory with the name name as the architecture (x86 or x64).

### appveyor.yml

Builds dependencies, then playd, then runs playd_tests.

All the required DLLs are zipped up with playd, playd_tests, and all licenses except for MPG123's.

All build zips are available via the Appveyor API, e.g. [here](https://ci.appveyor.com/api/projects/DinCahill/ury-playd/artifacts/playd__win64.zip?job=Platform: x64). If you push a tag to GitHub, Appveyor will create a release for the tag if it doesn't exist, and push the zips to it, named`playd_$TAG_NAME_win(32|64).zip`. Unfortunately, the zips of untagged builds have two underscores (e.g. playd__win64.zip), and I'm not sure how to fix it, but at least it's consistent.

The URLs to the dependencies are near the top of the BuildDeps function.

## TODO

~~I've not included the MPG123 license in the zip, because neither of the zips I'm downloading contains it. Since I'm modifying the MPG123 source with mpg123.patch, I probably do need to add the license. This patch does fix what looks like a proper bug, so I'll look into reporting it soon.~~ Fixed upstream. License added to zips.

#### Mandatory
- [x] Create a UniversityRadioYork account on Appveyor. There doesn't seem to be [any other way](http://help.appveyor.com/discussions/questions/1154-appveyor-account-for-github-organizations) to do it. @ChrisTheBaron pls weigh in.
- [x] Create a proper GitHub API token for pushing releases, encrypt it on Appveyor (Account -> Encrypt data), and replace `auth_token` in appveyor.yml.
- [x] Probably need MPG123 license (probably in the source).

#### Optional
- [x] Add more comments to appveyor.yml.
- [x] Add help text to WindowsBuilder.ps1.

## Other fixes included

- Fix a couple of `#ifndef`s I added to io.(cpp|hpp), which should've been `#ifdef`.
- Fix CMake policy warning [CMP0054](https://cmake.org/cmake/help/v3.6/policy/CMP0054.html).
- Turn off Link-Time Code Generation in MSVC, because all the libs we're using are shared.

### CODE REVIEW

**Comment by DinCahill** [comment]
The MPG123 maintainer kindly fixed the MPG123_EXPORT issue.
https://sourceforge.net/p/mpg123/bugs/239/

The latest release (1.23.6) also contains a def file, so all the gendef stuff
has been binned. Woop!

1.23.6 also changes `typedef long ssize_t` to `typedef ptrditt_t ssize_t`.
This slightly breaks my hack for getting MPG123 and libuv to live together,
and produces an extra warning on x64, but x64 has a lot of warnings about
size_t as well anyway, so these should be `FIX`.

Listening to MP3s still seems to work on x86 and x64!

**Comment by wlcx** [comment]
@DinCahill does https://www.appveyor.com/docs/team-setup/#setting-up-appveyor-account-for-github-organization change anything?

**Comment by DinCahill** [comment]
Use the standard Get-Help cmdlet to access the help text for WindowsBuilder.ps1.
`Get-Help .\WindowsBuilder.ps1 [ -full | -detailed | -examples ]`

**Comment by DinCahill** [comment]
I just need someone to enable pull requests for the webhook on the ury-playd repo. Then, I'll test PR building, and releasing.

Once that's done, I think it's ready to merge.

**Comment by DinCahill** [comment]
Ready to merge.

File: WindowsBuilder.ps1 [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+{
+	$python_version = & "$pypath" "--version" 2>&1
+	return $python_version -Match $vstring
+}
+
+# Check for Python 2.7, and throw an exception if not successful.
+function Check-Python ($vstring)
+{
+	if (Test-Path Env:\PYTHON)
+	{
```
Comment (LordAro) [inline WindowsBuilder.ps1]: Is this style of if {} else {} required in powershell? I'd prefer it if the opening/closing braces were on the same line, like the cpp code

File: WindowsBuilder.ps1 [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+if ($deps) {
+	BuildDeps $arch $depsdir $libdir $includedir $build
+}
+if ($playd) {
+	BuildPlayd $arch $archdir $build
+}
+if (!($deps -or $playd)) {
+	BuildDeps $arch $depsdir $libdir $includedir $build
+	BuildPlayd $arch $archdir $build
+}
```
Comment (LordAro) [inline WindowsBuilder.ps1]: trailing newline :>

Comment (LordAro) [inline WindowsBuilder.ps1]: I wonder if this file (and perhaps others) in root couldn't be put in a subdir, it'd be good to keep it as clean as possible

Comment (DinCahill) [inline WindowsBuilder.ps1]: 🎵 Whatever you want, whatever you like 🎵

Comment (DinCahill) [inline WindowsBuilder.ps1]: Will fix.

**Comment by DinCahill** [comment]
Gonna remove @CaptainHayashi's VS project and update README.VisualStudio.md.

**Pull request merged**